lib/redmica_s3/thumbnail_patch.rb

```diff
@@ -42,21 +44,23 @@ def generate(source, target, size, is_pdf = false)
             size_option = "#{size}x#{size}>"
             begin
               tempfile = MiniMagick::Utilities.tempfile(File.extname(source)) do |f| f.write(raw_data) end
-              convert_output =
-                if is_pdf
-                  MiniMagick::Tool::Convert.new do |cmd|
-                    cmd << "#{tempfile.to_path}[0]"
-                    cmd.thumbnail size_option
-                    cmd << 'png:-'
-                  end
-                else
-                  MiniMagick::Tool::Convert.new do |cmd|
-                    cmd << tempfile.to_path
-                    cmd.auto_orient
-                    cmd.thumbnail size_option
-                    cmd << '-'
-                  end
-                end
+              # Generate command
+              convert = MiniMagick::Tool::Convert.new
+              if is_pdf
+                convert << "#{tempfile.to_path}[0]"
+                convert.thumbnail size_option
+                convert << 'png:-'
+              else
+                convert << tempfile.to_path
+                convert.auto_orient
+                convert.thumbnail size_option
+                convert << '-'
+              end
+              # Execute command
+              convert_output = nil
+              Timeout.timeout(Redmine::Configuration['thumbnails_generation_timeout'].to_i) do
+                convert_output = convert.call
+              end
```

Comment:
I tested this pull request in my local environment and observed that the thumbnail generation process does not complete within the set timeout seconds.

Below is the Rails log when the timeout is set to 1 second on RedMica v3.0.1. The attached file used is a PDF with many objects, and it takes about 3 to 4 seconds to generate a thumbnail if the timeout is disabled.

```
Creating thumbnail timed out:
Command: convert /tmp/mini_magick20240628-1-9yuhsu.pdf[0] -thumbnail 200x200> png:-
Completed 404 Not Found in 4010ms (ActiveRecord: 0.8ms | Allocations: 15685)
```

I tried several times, but the result was the same. For reference, here is the result when testing the same PDF file with the redmica_s3 plugin version v2.0.0.

```
Completed 200 OK in 3951ms (Views: 3.0ms | ActiveRecord: 2.0ms | Allocations: 23266)
```

There is no change in the request time.

Additionally, I added debug prints to `AttachmentsController#thumbnail` and `Thumbnail#generate`, and here are the Rails log.

```
2024-06-28 05:21:02 +0000: [AttachmentsController#thumbnail] begin @attachment.thumbnail
2024-06-28 05:21:03 +0000: [Thumbnail#generate] begin Timeout.timeout
2024-06-28 05:21:03 +0000: [Thumbnail#generate] begin convert.call
...
2024-06-28 05:21:06 +0000: [Thumbnail#generate] rescue Timeout::Error
Creating thumbnail timed out:
Command: convert /tmp/mini_magick20240628-1-wuikx4.pdf[0] -thumbnail 200x200> png:-
2024-06-28 05:21:06 +0000: [AttachmentsController#thumbnail] end @attachment.thumbnail
2024-06-28 05:21:06 +0000: [AttachmentsController#thumbnail] rescue AttachmentsController#thumbnail
...
Completed 404 Not Found in 3598ms (ActiveRecord: 0.5ms | Allocations: 15961)
```
You can see that 3 seconds have already passed between `begin Timeout.timeout` and `rescue Timeout::Error`.

I am not sure, but it might be related to the fact that `MiniMagick::Tools::Convert` is also executed in `Thread`.
https://github.com/minimagick/minimagick/blob/fcd7600fe88e582431e7fef3c732accb8a42273e/lib/mini_magick/shell.rb#L43

Therefore, how about using MiniMagick's own timeout settings?
https://github.com/minimagick/minimagick/blob/fcd7600fe88e582431e7fef3c732accb8a42273e/lib/mini_magick/configuration.rb#L25-L30

```ruby
MIniMagick.timeout = 10
```

```diff
diff --git a/lib/redmica_s3/thumbnail_patch.rb b/lib/redmica_s3/thumbnail_patch.rb
index a46b557..d695481 100644
--- a/lib/redmica_s3/thumbnail_patch.rb
+++ b/lib/redmica_s3/thumbnail_patch.rb
@@ -1,5 +1,3 @@
-require 'timeout'
-
 module RedmicaS3
   module ThumbnailPatch
     extend ActiveSupport::Concern
@@ -44,6 +42,7 @@ module RedmicaS3
             size_option = "#{size}x#{size}>"
             begin
               tempfile = MiniMagick::Utilities.tempfile(File.extname(source)) do |f| f.write(raw_data) end
+
               # Generate command
               convert = MiniMagick::Tool::Convert.new
               if is_pdf
@@ -56,11 +55,12 @@ module RedmicaS3
                 convert.thumbnail size_option
                 convert << '-'
               end
+
               # Execute command
-              convert_output = nil
-              Timeout.timeout(Redmine::Configuration['thumbnails_generation_timeout'].to_i) do
-                convert_output = convert.call
+              convert_output = with_mini_magick_timeout(Redmine::Configuration['thumbnails_generation_timeout'].to_i) do
+                convert.call
               end
+
               img = MiniMagick::Image.read(convert_output)
 
               img_blob = img.to_blob
@@ -84,6 +84,18 @@ module RedmicaS3
           object.reload
           [object.metadata['digest'], object.get.body.read]
         end
+
+        def with_mini_magick_timeout(timeout_seconds)
+          return yield unless timeout_seconds > 0
+
+          begin
+            original_timeout = MiniMagick.timeout
+            MiniMagick.timeout = timeout_seconds
+            yield
+          ensure
+            MiniMagick.timeout = original_timeout
+          end
+        end
       end
     end
   end
```

For the full code, please refer to the following link.
https://github.com/hidakatsuya/redmica_s3/commit/9394675d76c87bdedbe5cb37755a3a2901404bdb

When tested with the same PDF file in a similar environment, the result was as follows:

```
Completed 404 Not Found in 1469ms (ActiveRecord: 0.6ms | Allocations: 15608)
```

A concern is that the timeout setting in MiniMagick is global. The above code resolves this by temporarily changing the timeout setting.

As per the following MiniMagick code, `MiniMagick::Tools::Convert` always references the `MiniMagick.timeout`.
https://github.com/minimagick/minimagick/blob/fcd7600fe88e582431e7fef3c732accb8a42273e/lib/mini_magick/shell.rb#L53

Reply:
@hidakatsuya Thank you for reviewing.

> I'm not sure, but it might be related to the fact that MiniMagick::Tools::Convert is also executed in Thread.
> https://github.com/minimagick/minimagick/blob/fcd7600fe88e582431e7fef3c732accb8a42273e/lib/mini_magick/shell.rb#L43

I agree. It seems so.

---

> As per the following MiniMagick code, MiniMagick::Tools::Convert always references the MiniMagick.timeout.
> https://github.com/minimagick/minimagick/blob/fcd7600fe88e582431e7fef3c732accb8a42273e/lib/mini_magick/shell.rb#L53

Created a monkey patch that override methods of the MiniMagick::Shell class so that you can pass in an arbitrary timeout value.
[49c6759](https://github.com/redmica/redmica_s3/pull/35/commits/49c67591265cc6446a8d52b231864447c862e1f1)